### PR TITLE
KFSPTS-29084 implement Jaggaer generate XML report

### PR DESCRIPTION
**Pull request by jhulslander**
There is a related non-prod-sql PR for this change
ConcurBatchUtilityService. getFileContents was deprecated.  The function was copied into LoadFileUtils,  This was done because this functionality is useful in multiple areas.  I didn't want to refactor everything that uses ConcurBatchUtilityService.getFileContents to avoid scope creep for this change.

I changed the LOG level in a catch block as it was just adding "cruff" to the job LOG file.  Other logging in the same area provide details needed if debugging is needed.

### CODE REVIEW

File: src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateSupplierXmlReportServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+            BodyMailMessage message = new BodyMailMessage();
+            message.setFromAddress(fromAddress);
+            String subject = reportWriterService.getTitle();
+            message.setSubject(subject);
+            message.getToAddresses().add(toAddress);
+            message.setMessage(body);
+    
+            boolean htmlMessage = false;
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("sendEmail, from address: " + fromAddress + "  to address: " + toAddress);
```
Comment (cah292) [inline src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateSupplierXmlReportServiceImpl.java:127]: There are several log statements in this method that need to update the mentioned method name: `sendEmail` --> `emailReport`.

Also, if desired, the debug statements could be parameterized and the call to `isDebugEnabled()` could be removed.

File: src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateSupplierXmlReportServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+                LOG.debug("sendEmail, the email subject: " + subject);
+                LOG.debug("sendEmail, the email budy: " + body);
```
Comment (cah292) [inline src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateSupplierXmlReportServiceImpl.java:129]: Minor typo in message: `budy` --> `body`

File: src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateSupplierXmlReportServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+            }
+            try {
+                emailService.sendMessage(message, htmlMessage);
+            } catch (Exception e) {
+                LOG.error("sendEmail, the email could not be sent", e);
+            }
+        } else {
+            LOG.info("emailReport, the JAGGAER_XML_REPORT_EMAIL parameter is empty, so not emailing the report");
```
Comment (cah292) [inline src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateSupplierXmlReportServiceImpl.java:137]: Is logging this message at info level okay, or should it be logged at warning or error level instead?

File: src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerUploadFileServiceImpl.java [hunk 4]
```diff
@@ -199,13 +199,9 @@ private URI buildSupplierUploadURI() {
         }
     }
 
-    private void generateResultsReport(List<JaggaerUploadFileResultsDTO> resultsList) {
-        /*
-         * A full report will be implemented on Jira KFSPTS-29084. Until then, this
-         * will produce a nice summary at the bottom of the log file
-         */
+    private void logUploadFileResusts(List<JaggaerUploadFileResultsDTO> resultsList) {
```
Comment (cah292) [inline src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerUploadFileServiceImpl.java:202]: Minor typo in method name: `...FileResusts` --> `...FileResults`

Also, is this method supposed to be converted into a report-generating one? Or has that work been deferred to a future ticket?

File: src/main/java/edu/cornell/kfs/sys/util/LoadFileUtils.java [hunk 5]
```diff
@@ -44,4 +45,20 @@ public static byte[] safelyLoadFileBytes(File file) {
...
+    
+    /**
+     * Returns the string contents of a file.  If the file can not be loaded, an empty string is returned.
+     * @param fullyQualifiedFileName
+     * @return
+     */
+    public static String safelyLoadFileString(String fullyQualifiedFileName) {
+        try {
+            byte[] fileByteArray = safelyLoadFileBytes(fullyQualifiedFileName);
+            String formattedString = new String(fileByteArray);
```
Comment (cah292) [inline src/main/java/edu/cornell/kfs/sys/util/LoadFileUtils.java:57]: I know the original method only used the single-arg String constructor, but you may want to consider updating this to use a constructor that also accepts a charset.

File: src/main/resources/edu/cornell/kfs/module/purap/cu-purap-resources.properties [hunk 6]
```diff
@@ -3,4 +3,11 @@ iwnt.note.create.reqs=Requisition edoc {0} created by {1} ({2})
 jaggaer.upload.xml.error.message=Message type '{0}' with a value of '{1}'
 jaggaer.upload.webservice.error=There was an error calling the Jaggaer upload service
 jaggaer.xml.website.error=Vendor number {0} has an invalid website URL of {1}
-jaggaer.xml.location.name.format={0} - {1}
\ No newline at end of file
+jaggaer.xml.location.name.format={0} - {1}
+jaggaer.xml.report.section.header=*********************** {0} **************************
+jaggaer.xml.report.section.footer=******************************************************
+jaggaer.xml.report.summary.file.name=Summary for {0}
+jaggaer.xml.report.summary.active.inactive.total=\t\t Total number of {0} vendors: {1}
```
Comment (cah292) [inline src/main/resources/edu/cornell/kfs/module/purap/cu-purap-resources.properties:10]: I see that the tab-prefixed messages also have a single space immediately after the tabs. Is that intentional, or should such lines be prefixed with only tabs or only spaces?

**Review by cah292 — CHANGES_REQUESTED** [review]
Changes look good overall, though I did notice some areas for further cleanup.

File: src/main/java/edu/cornell/kfs/concur/batch/service/ConcurBatchUtilityService.java [hunk 7]
```diff
@@ -139,9 +139,11 @@ public interface ConcurBatchUtilityService {
     
     /**
      * Returns the contents of a file as a String.
+     * @Deprecated use LoadFileUtils.safelyLoadFileString()
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/concur/batch/service/ConcurBatchUtilityService.java:142]: ok

File: src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateSupplierXmlReportServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+            reportWriterService.writeFormattedMessageLine(
+                    formatPropertyValue(CUPurapKeyConstants.JAGGAER_XML_REPORT_SUMMARY_ACTIVE_INACTIVE_NOTE_TOTAL,
+                            INACTIVE, WITHOUT, String.valueOf(filterDetailDto(dto.getInactiveVendors(), false).size())));
+
+            dtoCount++;
+        }
+        reportWriterService.writeSubTitle(formatPropertyValue(CUPurapKeyConstants.JAGGAER_XML_REPORT_SECTION_FOOTER));
+    }
+
+    private String formatPropertyValue(String propertyName, String... args) {
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateSupplierXmlReportServiceImpl.java:78]: ok

File: src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateSupplierXmlReportServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+        String propertyValue = configurationService.getPropertyValueAsString(propertyName);
+        return MessageFormat.format(propertyValue, args);
+    }
+
+    private List<JaggaerUploadSupplierVendorDetailDto> filterDetailDto(List<JaggaerUploadSupplierVendorDetailDto> dtos, boolean hasNotes) {
+        return dtos.stream()
+                .filter(dto -> dto.hasNotes() == hasNotes)
+                .collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateSupplierXmlReportServiceImpl.java:86]: ok

File: src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateSupplierXmlReportServiceImpl.java (hunk 1 above)
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateSupplierXmlReportServiceImpl.java:127]: ok

File: src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateSupplierXmlReportServiceImpl.java (hunk 2 above)
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateSupplierXmlReportServiceImpl.java:129]: ok

File: src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateSupplierXmlReportServiceImpl.java (hunk 3 above)
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateSupplierXmlReportServiceImpl.java:137]: good point, I will switch to warn

File: src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerUploadFileServiceImpl.java (hunk 4 above)
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerUploadFileServiceImpl.java:202]: I switched the name to logUploadFileResults

The functionals currently believe they don't need this particular report.  Jaggaer generates a report internally if there are errors on the upload.  We should continue to log the results to the job run log to aid in trouble shooting if there are upload errors.

File: src/main/java/edu/cornell/kfs/sys/util/LoadFileUtils.java (hunk 5 above)
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/sys/util/LoadFileUtils.java:57]: ok

**Comment by jhulslander** [comment]
this is ready for review again

File: src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerUploadFileServiceImpl.java (hunk 4 above)
Comment (cah292) [reply src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerUploadFileServiceImpl.java]: Ok, thanks for the explanation.

File: src/main/resources/edu/cornell/kfs/module/purap/cu-purap-resources.properties (hunk 6 above)
Comment (cah292) [reply src/main/resources/edu/cornell/kfs/module/purap/cu-purap-resources.properties]: Thanks for the clarification. I'm fine with the tabs and spaces being left in as-is.

File: src/main/java/edu/cornell/kfs/sys/util/LoadFileUtils.java [hunk 10]
```diff
@@ -44,4 +47,31 @@ public static byte[] safelyLoadFileBytes(File file) {
         }
         return safelyLoadFileBytes(file.getAbsolutePath());
     }
+    
+    /**
+     * Returns the string contents of a file.  If the file can not be loaded, an empty string is returned.
+     * Uses a default UTF_*
```
Comment (cah292) [inline src/main/java/edu/cornell/kfs/sys/util/LoadFileUtils.java:53]: Did you mean to say "UTF_8" in this comment?

**Review by cah292 — CHANGES_REQUESTED** [review]
I noticed a few more items that may need cleanup, but otherwise the changes look good.

File: src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateSupplierXmlReportServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+        int dtoCount = 0;
+        for (JaggaerUploadSupplierXmlFileDetailsDto dto : xmlFileDtos) {
+            if (dtoCount > 0) {
+                reportWriterService.writeNewLines(2);
+            }
+            reportWriterService.writeFormattedMessageLine(configurationService.getPropertyValueAsString(
+                    CUPurapKeyConstants.JAGGAER_XML_REPORT_SUMMARY_FILE_NAME), dto.getXmlFileName());
+
+            reportWriterService.writeFormattedMessageLine(
+                    configurationService.getPropertyValueAsString(CUPurapKeyConstants.JAGGAER_XML_REPORT_SUMMARY_ACTIVE_INACTIVE_TOTAL), ACTIVE,
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateSupplierXmlReportServiceImpl.java:53]: ok

File: src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateSupplierXmlReportServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+                    configurationService.getPropertyValueAsString(CUPurapKeyConstants.JAGGAER_XML_REPORT_SUMMARY_ACTIVE_INACTIVE_NOTE_TOTAL),
+                            INACTIVE, WITHOUT, filterDetailDtoWithoutNotes(dto.getInactiveVendors()).size());
+
+            dtoCount++;
+        }
+        reportWriterService.writeSubTitle(configurationService.getPropertyValueAsString(CUPurapKeyConstants.JAGGAER_XML_REPORT_SECTION_FOOTER));
+    }
+    
+   
+    private String writeSubTitleValue(String propertyName, Object... args) {
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateSupplierXmlReportServiceImpl.java:78]: ok

File: src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateSupplierXmlReportServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+            
+            detail.getNotes().stream().forEach(note -> 
+                reportWriterService.writeFormattedMessageLine(
+                        configurationService.getPropertyValueAsString(CUPurapKeyConstants.JAGGAER_XML_REPORT_DETAIL_NOTE_LINE), note));
+        });
+    }
+    
+    private void emailReport() {
+        final String toAddress = findReportToAddress();
+        final String fromAddress = toAddress;
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateSupplierXmlReportServiceImpl.java:122]: oh, thank you, I forgot about that, I am going to switch the from address

File: src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateSupplierXmlReportServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+    
+            boolean htmlMessage = false;
+            LOG.debug("emailReport, from address: {}  to address: {}", fromAddress, toAddress);
+            LOG.debug("emailReport, the email subject: {}",  subject);
+            LOG.debug("emailReport, the email body: {}",  body);
+                
+            try {
+                emailService.sendMessage(message, htmlMessage);
+            } catch (Exception e) {
+                LOG.error("sendEmail, the email could not be sent", e);
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateSupplierXmlReportServiceImpl.java:141]: ok

File: src/main/java/edu/cornell/kfs/sys/util/LoadFileUtils.java (hunk 10 above)
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/sys/util/LoadFileUtils.java:53]: yes, will fix

**Comment by jhulslander** [comment]
this is ready for review again

**Review by cah292 — COMMENTED** [review]
Latest changes look good. I see that the unit tests supposedly passed, but due to technical issues with the instance that runs the tests, I'm wondering if it would be best to wait for that to be resolved before I merge this, just to confirm that the tests did indeed run.

**Comment by cah292** [comment]
As per the recent team meeting, we're confident that the tests executed and passed as expected. I will proceed with the merge.

**Pull request merged**